### PR TITLE
Replace csrf plugin with Queensland Government solution

### PR DESCRIPTION
**Pull request by anotheredward**
Our current CSRF solution requires server-side state and doesn't actually protect against CSRF (it does make it very difficult, but not impossible).

This PR replaces it with the approach the Queensland Government have taken which both does protect appropriately and doesn't require server-side state.

Ok, thanks for the help @ThrawnCA , I've managed to get it up and working, the following potential tasks remain:

### CODE REVIEW

File: ckanext/security/anti_csrf.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    return token
+
+def get_response_token():
+    """Retrieve the token to be injected into pages.
+
+    This will be retrieved from the 'token' cookie, if it exists and is fresh.
+    If not, a new token will be generated and a new cookie set.
+    """
+    # ensure that the same token is used when a page is assembled from pieces
+    if request.environ['webob.adhoc_attrs'].has_key('response_token'):
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/security/anti_csrf.py]: Not sure; I've never seen it happen.

File: ckanext/security/anti_csrf.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+"""The format of the token HTML field.
+"""
+HEX_PATTERN=re.compile(r'^[0-9a-z]+$')
+TOKEN_PATTERN = r'<input type="hidden" name="' + TOKEN_FIELD_NAME + '" value="{token}"/>'
+TOKEN_SEARCH_PATTERN = re.compile(TOKEN_PATTERN.format(token=r'([0-9a-f]+)'))
+API_URL = re.compile(r'^/api\b.*')
+
+def is_logged_in():
+    return True
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/security/anti_csrf.py]: It's actually usable, albeit inefficient, as it stands.

The original implementation checks for the presence of the `auth_tkt` cookie, but this doesn't protect against Login CSRF or brute-force attempts, and apparently there's a migration in the works to replace auth_tkt with Beaker. Probably the ideal would be to check for both old and new cookies, and also check the URL to see if it's the login page even if using GET.

File: ckanext/security/anti_csrf.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+def anti_csrf_before(obj, action, **params):
+    if not is_request_exempt() and get_cookie_token() != get_post_token():
+        csrf_fail("Could not match session token with form token")
+
+    RAW_BEFORE(obj, action)
+
+def csrf_fail(message):
+    from flask import abort
+    LOG.error(message)
+    abort(403, "Your form submission could not be validated")
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/security/anti_csrf.py]: Open to suggestions here.

File: ckanext/security/anti_csrf.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        request.token = postTokens[0]
+
+    # drop token from request so it doesn't populate resource extras
+    del request.POST[TOKEN_FIELD_NAME]
+
+    return request.token
+
+def intercept_csrf():
+    base.render_jinja2 = anti_csrf_render_jinja2
+    base.BaseController.__before__ = anti_csrf_before
```
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/security/anti_csrf.py]: Just based on https://docs.ckan.org/en/2.8/extensions/plugin-interfaces.html, it looks like "No". There are hooks for specific form types (dataset, resource, etc) which perhaps could be used to insert tokens, but there's not a general hook for pre- and post-rendering.

Perhaps it's worth suggesting at https://github.com/ckan/ideas-and-roadmap/ ?

**Comment by ThrawnCA** [comment]
This change is not backwards-compatible; any installations that are patching the CKAN middleware will get errors, because the middleware module is no longer available.

Also, the README should be updated.

**Comment by ebuckley** [comment]
Just an update with where this is at right now, I'm picking up closing off this issue and integrating the double cookie CSRF protection.

I have noticed a couple issues with the current implementation which need to be worked out before we can merge this.

1. The `is_logged_in()` method checks for the `auth_tkt` cookie, but this does not exist when using this plugin as we use Beaker for session tracking (See who.ini config that is required for this plugin)
2. The CSRF check code does not get executed for the auth routes (login/logout)
3. The origin/referrer header check needs to have webserver config, where we would prefer to have this check in the application.

I think we can cover these issues with an implementation of the CSRF check from our current Middleware class. I have a branch for these changes, but still need a bit of refinement before I'm happy with the implementation.

**Comment by onkartibe** [comment]
LGTM

File: ckanext/security/anti_csrf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThrawnCA) [reply ckanext/security/anti_csrf.py]: Perhaps the `abort` function from the [plugin toolkit](https://docs.ckan.org/en/latest/extensions/plugins-toolkit.html)?

**Comment by camfindlay** [comment]
Let's look to merge this and make a new SEMVER release of this module. Question @ebuckley have we tagged the current version so those looking to upgrade get the hint that this is a more major release?

**Comment by ThrawnCA** [comment]
@camfindlay To properly follow SEMVER requires a public API. I suppose the closest thing to that would be the configuration settings? And the middleware patching.

Since this does change how the extension interacts with middleware - ie it no longer needs the patch - but it should be compatible with existing configuration, I suggest updating the minor version to 1.1.0.

**Comment by camfindlay** [comment]
Minor release seems reasonable - it's backwards compatible (even though just upgrading in place might be breaking if you dont tidy up). 1.1.0 it is.

**Comment by ebuckley** [comment]
>Since this does change how the extension interacts with middleware - ie it no longer needs the patch - but it should be compatible with existing configuration, I suggest updating the minor version to 1.1.0.

FWIW the change here will still requires the middle-ware patch. The middleware is required because there are several routes that were not triggering the previous method of hooking into the request/response 

All that said, I'm happy to go with a minor release to 1.1.0

I will merge this PR and action that update now.

**Pull request merged**